File: tests/test_command_prompt_piping.py
# pyright: standard

from pathlib import Path

from pytest_mock import MockerFixture
from typer.testing import CliRunner

from aico.main import app
from aico.models import TokenUsage
from tests import helpers

runner = CliRunner()


def test_gen_successful_diff_piped(tmp_path: Path, mocker: MockerFixture) -> None:
    """
    Tests the 'Strict Contract' for `aico gen`: successful diffs are printed to stdout.
    """
    # GIVEN a session with a file and a mocked LLM returning a valid diff
    llm_response = "File: file.py\n<<<<<<< SEARCH\nold content\n=======\nnew content\n>>>>>>> REPLACE"
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        # For these tests, we are always in a non-TTY (piped) environment
        mocker.patch("aico.commands.prompt.is_terminal", return_value=False)
        mocker.patch("aico.llm.executor.is_terminal", return_value=False)

        usage = TokenUsage(prompt_tokens=10, completion_tokens=10, total_tokens=20, cost=0.005)

        helpers.setup_test_session_and_llm(
            runner,
            app,
            Path(td),
            mocker,
            llm_response,
            context_files={"file.py": "old content"},
            usage=usage,
        )

        # WHEN `aico gen` is run
        result = runner.invoke(app, ["gen", "a prompt"])

        # THEN the command succeeds
        assert result.exit_code == 0
        # AND stdout contains ONLY the clean, unified diff
        expected_diff = (
            "--- a/file.py\n+++ b/file.py\n@@ -1 +1 @@\n-old content\n\\ No newline at end of file\n+new content\n"
        )
        assert result.stdout == expected_diff
        # AND stderr contains cost info but no warnings
        assert "Cost:" in result.stderr
        assert "Warning" not in result.stderr


def test_gen_failing_diff_piped(tmp_path: Path, mocker: MockerFixture) -> None:
    """
    Tests the 'Strict Contract' for `aico gen`: failing diffs result in an empty stdout.
    """
    # GIVEN a session with a file and a mocked LLM returning a failing diff
    llm_response = "File: file.py\n<<<<<<< SEARCH\ncontent not found\n=======\nnew content\n>>>>>>> REPLACE"
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        mocker.patch("aico.commands.prompt.is_terminal", return_value=False)
        mocker.patch("aico.llm.executor.is_terminal", return_value=False)
        usage = TokenUsage(prompt_tokens=10, completion_tokens=10, total_tokens=20, cost=0.005)

        helpers.setup_test_session_and_llm(
            runner,
            app,
            Path(td),
            mocker,
            llm_response,
            context_files={"file.py": "old content"},
            usage=usage,
        )

        # WHEN `aico gen` is run
        result = runner.invoke(app, ["gen", "a prompt"])

        # THEN the command succeeds
        assert result.exit_code == 0
        # AND stdout is an empty string
        assert result.stdout == ""
        # AND stderr contains the warning about the patch failure
        assert "Warnings:" in result.stderr
        assert "The SEARCH block from the AI could not be found" in result.stderr


def test_ask_conversational_text_piped(tmp_path: Path, mocker: MockerFixture) -> None:
    """
    Tests the 'Flexible Contract' for `aico ask`: conversational text is printed to stdout.
    """
    # GIVEN a session and an LLM returning plain text
    llm_response = "This is a simple conversational response."
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        mocker.patch("aico.commands.prompt.is_terminal", return_value=False)
        mocker.patch("aico.llm.executor.is_terminal", return_value=False)
        usage = TokenUsage(prompt_tokens=10, completion_tokens=10, total_tokens=20, cost=0.005)

        helpers.setup_test_session_and_llm(runner, app, Path(td), mocker, llm_response, usage=usage)

        # WHEN `aico ask` is run
        result = runner.invoke(app, ["ask", "a prompt"])

        # THEN the command succeeds
        assert result.exit_code == 0
        # AND stdout contains the conversational text
        assert result.stdout == llm_response


def test_ask_successful_diff_piped(tmp_path: Path, mocker: MockerFixture) -> None:
    """
    Tests the 'Flexible Contract' for `aico ask`: a successful diff is prioritized and printed to stdout.
    """
    # GIVEN a session, a file, and an LLM returning a valid diff
    llm_response = "File: file.py\n<<<<<<< SEARCH\nold content\n=======\nnew content\n>>>>>>> REPLACE"
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        mocker.patch("aico.commands.prompt.is_terminal", return_value=False)
        mocker.patch("aico.llm.executor.is_terminal", return_value=False)
        usage = TokenUsage(prompt_tokens=10, completion_tokens=10, total_tokens=20, cost=0.005)

        helpers.setup_test_session_and_llm(
            runner,
            app,
            Path(td),
            mocker,
            llm_response,
            context_files={"file.py": "old content"},
            usage=usage,
        )

        # WHEN `aico ask` is run
        result = runner.invoke(app, ["ask", "a prompt"])

        # THEN the command succeeds
        assert result.exit_code == 0
        # AND stdout contains the clean, unified diff
        expected_diff = (
            "--- a/file.py\n+++ b/file.py\n@@ -1 +1 @@\n-old content\n\\ No newline at end of file\n+new content\n"
        )
        assert result.stdout == expected_diff


def test_ask_failing_diff_piped(tmp_path: Path, mocker: MockerFixture) -> None:
    """
    Tests the 'Flexible Contract' for `aico ask`: a failing diff falls back to printing the `display_content` to stdout.
    """
    # GIVEN a session, a file, and an LLM returning a failing diff
    llm_response = "File: file.py\n<<<<<<< SEARCH\ncontent not found\n=======\nnew content\n>>>>>>> REPLACE"
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        mocker.patch("aico.commands.prompt.is_terminal", return_value=False)
        mocker.patch("aico.llm.executor.is_terminal", return_value=False)
        usage = TokenUsage(prompt_tokens=10, completion_tokens=10, total_tokens=20, cost=0.005)

        helpers.setup_test_session_and_llm(
            runner,
            app,
            Path(td),
            mocker,
            llm_response,
            context_files={"file.py": "old content"},
            usage=usage,
        )

        # WHEN `aico ask` is run
        result = runner.invoke(app, ["ask", "a prompt"])

        # THEN the command succeeds
        assert result.exit_code == 0
        # AND stdout contains the fallback display_content with the rich error
        assert "The SEARCH block from the AI could not be found" in result.stdout
        assert "```diff" not in result.stdout  # The raw block is shown, not a diff
        assert "<<<<<<< SEARCH" in result.stdout
        # AND stderr contains the warning as well
        assert "Warnings:" in result.stderr
        assert "The SEARCH block from the AI could not be found" in result.stderr

File: tests/test_command_prompt_piping.rs
mod common;
use assert_cmd::cargo::cargo_bin_cmd;
use mockito::{Matcher, Server};
use serde_json::json;
use std::fs;
use tempfile::tempdir;

#[tokio::test]
async fn test_prompt_input_scenarios_both_cli_and_piped() {
    let temp = tempdir().unwrap();
    let root = temp.path();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .arg("--model")
        .arg("openai/test-model")
        .assert()
        .success();

    let mut server = Server::new_async().await;

    let expected_body = json!({
        "model": "test-model",
        "messages": [
            { "role": "system", "content": aico::consts::DEFAULT_SYSTEM_PROMPT },
            { "role": "user", "content": aico::consts::ALIGNMENT_CONVERSATION_USER },
            { "role": "assistant", "content": aico::consts::ALIGNMENT_CONVERSATION_ASSISTANT },
            {
                "role": "user",
                "content": "<stdin_content>\npipe in\n</stdin_content>\n<prompt>\ncli arg\n</prompt>"
            }
        ],
        "stream": true,
        "stream_options": { "include_usage": true }
    });

    let mock = server
        .mock("POST", "/chat/completions")
        .match_body(Matcher::Json(expected_body))
        .with_status(200)
        .with_header("content-type", "text/event-stream")
        .with_body("data: {}\n\ndata: [DONE]\n\n")
        .create_async()
        .await;

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("OPENAI_API_KEY", "sk-test")
        .env("OPENAI_BASE_URL", server.url())
        .args(["ask", "cli arg"])
        .write_stdin("pipe in")
        .assert()
        .success();

    mock.assert_async().await;

    // Verify history saved both components
    let history_path = root.join(".aico/history/0.jsonl");
    let content = fs::read_to_string(history_path).unwrap();
    assert!(content.contains("\"content\":\"cli arg\""));
    assert!(content.contains("\"piped_content\":\"pipe in\""));
}

#[tokio::test]
async fn test_ask_successful_diff_piped_flexible_contract() {
    let temp = tempdir().unwrap();
    let root = temp.path();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .arg("--model")
        .arg("openai/test-model")
        .assert()
        .success();

    let file_path = root.join("app.py");
    fs::write(&file_path, "old content").unwrap();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["add", "app.py"])
        .assert()
        .success();

    let mut server = Server::new_async().await;
    let llm_diff = "I updated the file.\n\nFile: app.py\n<<<<<<< SEARCH\nold content\n=======\nnew content\n>>>>>>> REPLACE";

    let mock = server
        .mock("POST", "/chat/completions")
        .with_status(200)
        .with_header("content-type", "text/event-stream")
        .with_body(format!(
            "data: {}\n\ndata: [DONE]\n\n",
            json!({"choices": [{"delta": {"content": llm_diff}}]})
        ))
        .create_async()
        .await;

    // WHEN running ask in piped mode
    let assert = cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("OPENAI_API_KEY", "sk-test")
        .env("OPENAI_BASE_URL", server.url())
        .args(["ask", "change"])
        .assert()
        .success();

    // THEN it converts to Unified Diff for stdout in preference to text
    let stdout = String::from_utf8(assert.get_output().stdout.clone()).unwrap();
    assert!(stdout.contains("--- a/app.py"));
    assert!(stdout.contains("-old content"));
    assert!(stdout.contains("+new content"));
    // Conversational garbage is stripped from stdout in piping mode if a diff is present
    assert!(!stdout.contains("I updated the file"));

    mock.assert_async().await;
}

#[tokio::test]
async fn test_gen_command_generates_diff_non_tty() {
    let temp = tempdir().unwrap();
    let root = temp.path();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .arg("--model")
        .arg("openai/test-model")
        .assert()
        .success();

    let file_path = root.join("file.py");
    fs::write(&file_path, "old line").unwrap();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["add", "file.py"])
        .assert()
        .success();

    let mut server = Server::new_async().await;
    let llm_diff = "Update!\n\nFile: file.py\n<<<<<<< SEARCH\nold line\n=======\nnew line\n>>>>>>> REPLACE\n\nLater!";

    let mock = server
        .mock("POST", "/chat/completions")
        .with_status(200)
        .with_header("content-type", "text/event-stream")
        .with_body(format!(
            "data: {}\n\ndata: [DONE]\n\n",
            json!({"choices": [{"delta": {"content": llm_diff}}]})
        ))
        .create_async()
        .await;

    let assert = cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("OPENAI_API_KEY", "sk-test")
        .env("OPENAI_BASE_URL", server.url())
        .args(["gen", "make change"])
        .assert()
        .success();

    let output = assert.get_output();
    let stdout = String::from_utf8(output.stdout.clone()).unwrap();

    // Strict contract: ONLY diff in stdout
    assert!(stdout.contains("--- a/file.py"));
    assert!(stdout.contains("+new line"));
    assert!(!stdout.contains("Update!"));
    assert!(!stdout.contains("Later!"));

    mock.assert_async().await;
}

#[tokio::test]
async fn test_prompt_with_history_reconstructs_piped_content_correctly() {
    let temp = tempdir().unwrap();
    let root = temp.path();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .arg("--model")
        .arg("openai/test-model")
        .assert()
        .success();

    let mut server = Server::new_async().await;

    // First turn with piped input
    let mock1 = server
        .mock("POST", "/chat/completions")
        .with_status(200)
        .with_header("content-type", "text/event-stream")
        .with_body("data: {\"choices\":[{\"delta\":{\"content\":\"resp 1\"}}]}\n\ndata: [DONE]\n\n")
        .create_async()
        .await;

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("OPENAI_API_KEY", "sk-test")
        .env("OPENAI_BASE_URL", server.url())
        .args(["ask", "first cli"])
        .write_stdin("first pipe")
        .assert()
        .success();
    mock1.assert_async().await;

    // Second turn: Verify the previous turn is reconstructed with XML
    let expected_body = json!({
        "model": "test-model",
        "messages": [
            { "role": "system", "content": aico::consts::DEFAULT_SYSTEM_PROMPT },
            {
                "role": "user",
                "content": "<stdin_content>\nfirst pipe\n</stdin_content>\n<prompt>\nfirst cli\n</prompt>"
            },
            { "role": "assistant", "content": "resp 1" },
            { "role": "user", "content": aico::consts::ALIGNMENT_CONVERSATION_USER },
            { "role": "assistant", "content": aico::consts::ALIGNMENT_CONVERSATION_ASSISTANT },
            { "role": "user", "content": "second prompt" }
        ],
        "stream": true,
        "stream_options": { "include_usage": true }
    });

    let mock2 = server
        .mock("POST", "/chat/completions")
        .match_body(Matcher::Json(expected_body))
        .with_status(200)
        .with_header("content-type", "text/event-stream")
        .with_body("data: {}\n\ndata: [DONE]\n\n")
        .create_async()
        .await;

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("OPENAI_API_KEY", "sk-test")
        .env("OPENAI_BASE_URL", server.url())
        .args(["ask", "second prompt"])
        .assert()
        .success();

    mock2.assert_async().await;
}

#[tokio::test]
async fn test_ask_conversational_text_piped() {
    let temp = tempdir().unwrap();
    let root = temp.path();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .arg("--model")
        .arg("openai/test-model")
        .assert()
        .success();

    let mut server = Server::new_async().await;
    let llm_text = "This is a conversational response.";

    let body = json!({
        "choices": [{"delta": {"content": llm_text}}]
    });
    let mock = server
        .mock("POST", "/chat/completions")
        .with_status(200)
        .with_header("content-type", "text/event-stream")
        .with_body(format!("data: {}\n\ndata: [DONE]\n\n", body))
        .create_async()
        .await;

    let output = cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("OPENAI_API_KEY", "sk-test")
        .env("OPENAI_BASE_URL", server.url())
        .args(["ask", "say something"])
        .assert()
        .success()
        .get_output()
        .stdout
        .clone();

    assert_eq!(String::from_utf8(output).unwrap().trim(), llm_text);
    mock.assert_async().await;
}

#[tokio::test]
async fn test_ask_failing_diff_piped() {
    let temp = tempdir().unwrap();
    let root = temp.path();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .arg("--model")
        .arg("openai/test-model")
        .assert()
        .success();

    let mut server = Server::new_async().await;
    // Search block doesn't exist on disk
    let llm_response =
        "File: non-existent.py\n<<<<<<< SEARCH\nmissing\n=======\nnew\n>>>>>>> REPLACE";

    let body = json!({
        "choices": [{"delta": {"content": llm_response}}]
    });
    let mock = server
        .mock("POST", "/chat/completions")
        .with_status(200)
        .with_header("content-type", "text/event-stream")
        .with_body(format!("data: {}\n\ndata: [DONE]\n\n", body))
        .create_async()
        .await;

    let assert = cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("OPENAI_API_KEY", "sk-test")
        .env("OPENAI_BASE_URL", server.url())
        .args(["ask", "change file"])
        .assert()
        .success();

    let output = assert.get_output();
    let stdout = String::from_utf8(output.stdout.clone()).unwrap();
    let stderr = String::from_utf8(output.stderr.clone()).unwrap();

    // Flexible contract: If diff fails in ASK mode, print raw text/warnings to user
    assert!(stdout.contains("<<<<<<< SEARCH"));
    assert!(stderr.contains("Warnings:"));
    assert!(stderr.contains("File 'non-existent.py' from the AI does not match"));

    mock.assert_async().await;
}

#[tokio::test]
async fn test_gen_failing_diff_piped() {
    let temp = tempdir().unwrap();
    let root = temp.path();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .arg("--model")
        .arg("openai/test-model")
        .assert()
        .success();

    let mut server = Server::new_async().await;
    let llm_response =
        "File: non-existent.py\n<<<<<<< SEARCH\nmissing\n=======\nnew\n>>>>>>> REPLACE";

    let body = json!({
        "choices": [{"delta": {"content": llm_response}}]
    });
    let mock = server
        .mock("POST", "/chat/completions")
        .with_status(200)
        .with_header("content-type", "text/event-stream")
        .with_body(format!("data: {}\n\ndata: [DONE]\n\n", body))
        .create_async()
        .await;

    let assert = cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("OPENAI_API_KEY", "sk-test")
        .env("OPENAI_BASE_URL", server.url())
        .args(["gen", "change file"])
        .assert()
        .success();

    let output = assert.get_output();
    // Strict contract: If diff fails in GEN mode, stdout must be EMPTY to avoid corrupting patch pipes
    assert!(output.stdout.is_empty());
    assert!(
        String::from_utf8(output.stderr.clone())
            .unwrap()
            .contains("Warnings:")
    );

    mock.assert_async().await;
}
